### ISSUE TITLE
SyncRouterState delay

### ISSUE THREAD

**Issue by adriaanwm** [inputlogic/elements#135]
Since updating to the new Router component, I needed the router values in state, so I used `SyncRouterState`. Unfortunately this has lead to a delay between the route values being in state and the component that is being rendered. For example if I have a user page `/users/:userId` and get the userId with useMapped state in the `User` component, the value will be undefined first, adding extra if logic (which is especially annoying when using hooks since you can't have any hooks after the if statement)

**Comment by staydecent** [comment]
@adriaanwm The route args are passed as props still: https://github.com/inputlogic/elements/blob/master/components/router/router.js#L236

```
function User ({ userId }) { ... }
```

**Comment by adriaanwm** [comment]
@staydecent this only helps if it isn't a nested component, unless I start passing props multiple levels deep

**Comment by staydecent** [comment]
Mmm, aren't most API calls related to current URL made in the "page" component though? And, if not, "dependent fetching" pattern should work? https://github.com/inputlogic/elements/tree/master/components/use-request#dependent-fetching

We could look into delaying rendering of the found route component in Router, until after the context/react-state is updated. But this would add significant complexity to Router.

**Comment by adriaanwm** [comment]
hmm, yeah you're right dependent fetching will solve the problem wherever the value is being used for a request. I feel like this is a bandaid soln in that every hook we build will have to accommodate undefined values otherwise we have to break things into more and more components in order to have `if` logic. 
Agree that waiting for context will add complexity and probably more bugs to Router. 
I'm getting the sense we're gradually moving away from the store being the source of app state, now it's spread between component state, context, and the store. My preference would be to keep pushing things to global store rather than away from it.

**Comment by staydecent** [comment]
Hmm, I don't know if I'd call it a bandaid solution. Outside of React, functions often need to account for null or bad values. I feel this is inevitable as a function is used in a wider array of situations.

The goal was to remove non-react dependencies from elements. Global store/app state is now explicitly managed via the pwa-starter code. This also allows us to move to any other state management library without having to untangle it from elements.

If it helps, I also don't think of Context as a state management layer, but rather just a transport/communication layer 😅 . State is only managed via react useState hook, but is communicated between components via Context.

Anywhoodle, I agree the delay in syncing is not ideal. Worth keeping this open to come up with ideas. I have one such I can test out today.

**Comment by adriaanwm** [comment]
I do like the decoupling dependencies (for me this includes react ones like context) from elements.
My preference would be to have a setup layer where you connect components from elements to your projects way of handling things like state. Something like this:

```
import { Router as RouterElement, match: matchBase } from '@app-elements/router'
import { store } from '/store'
import allRoutes from '/routes'

// match returns an object like {name: 'MatchedName', args: ...}
export const match = href => allRoutes.find(r => matchBase(r, href))

// RouterElement is only responsible for returning a component named `name`
export const Router = ({routes}) => {
    const name = useMappedState(store, W.path(['route', 'name']))
    return RouterElement({name, routes})
}

export const Link = ({href, ...props}) => {
   const onClickLink = (ev) => {
      ev.preventDefault()
      store.setState({route: match(href)})
   }
   return <a
      href={href}
      onClick={onClickLink}
   />
}
```

Store initial state would have to be a bit smarter:

```
import { match } from '/util/router'

const store = createStore({
    route: match(window.location.pathname)
})
```

This is brainstorming more than anything, I'd be happy to make this functional if you like this direction @staydecent

**Comment by staydecent** [comment]
I wonder if this is taking it too far?

I don't follow decoupling Context, when React *is* a dependency? These elements are all 100% intended to be used in React/Preact projects, with no allusion to being compatible with Vue, Angular, etc. Context is a decent API, as it allows using `props` without being limited to a passing props through a top-down tree.

This solution also puts a heavier load on the end-user of the elements. And, even requiring the end-user has a state layer they want to integrate with at all.

I think the timing issue can be solved, with either a nested internal component that renders the found route based on context, rather than directly in the Router component, or by exposing a new callback as a prop on Router, that can be invoked at the same point that `setRoute` is invoked internally.

Are there route args that you use other than to define an API request URL? How often do you access route args *not* within the matched route Component?

**Comment by adriaanwm** [comment]
1. Since we want the route values in store, handling the value with both Context & store increases js size & number of moving parts. The Router I'm suggesting is bordering on completely functional, it takes name and components and returns the matching component. We could include an example of how to use the router with context. There is also no race condition between the path and the route args in state because we are setting those values at time of path change.

2. I'm curious what criteria to look at to decide if an API is good or not. By my measures (namely explicit, functional, does not break basic coding standards, simple) hooks and Context are not good APIs. Especially hooks. The idea of only being able to call functions in a certain order without if statements would rightfully be turned down for any library other than React, I don't know why we or anyone else has made an exception. (hooks have been giving me grief so putting this here even though it's a bit of a tangent)

Also, avoiding dependencies should come down to more than what's included in an existing package (and for similar reasoning to the packages one). More moving parts, more complex, more lines of code etc.

"Are there route args that you use other than to define an API request URL? How often do you access route args not within the matched route Component?"
I don't think it matters how often route args are used in a nested component or in what way they're used. If we try to only handle the situations we can think of we'll be missing a lot.

I don't mean to brush aside the intention of having packages that you can install and they just work, I agree that should be one of our goals. I just think that shouldn't win out over everything else (especially how useful the package is for how we use it)

**Comment by staydecent** [comment]
Interesting, I think we have fundamental difference of opinion on this.

1.  The current router clocks in at 10kb unminified. Reach Router is 20kb. Preact Router is 9kb but doesn't support nesting. If we want to have a nice website promoting that people use these elements (which I think we should), they should work with React out of the box. Syncing the router state doesn't complicate how the Router works, there is no additional "moving parts" introduced to the router. There is additional moving parts added to the host project. But, that's the cost of being able to swap state management at the project level, no?

2. Really? I think Context is a good API. It's functional, explicit, and only works by reference rather than a descriptive string lookup. The API is:

```
createContext => create a new context
Context.Provider => Provide the value for the context
Context.Consumer => Consume the value for the context
```

As for hooks, 100% agree the limitation of them needing to be called any every render (so not behind an if statement) is weird. However, beyond that, they are an improvement on Class based syntax. I also think you would find gotcha's unique to each front-end framework. I don't think that's good by any means, but it is the reality of front-end development.

> I don't think it matters how often route args are used in a nested component or in what way they're used. If we try to only handle the situations we can think of we'll be missing a lot.

Absolutely, it matters. How else can we weigh the costs without understanding the real-life scenarios being encountered? For instance, in your project, are there 1, 10, 20 components relying on route args that are not the route component themselves? In other words, if this is an issue on 100% of the routes you have defined in your project, then there's more weight or merit to addressing this issue in a particular manner.

And, as I can't think of another scenario of needing the route arg to pass to a useRequest hook, I don't have the context to understand other current needs that need solving.

How can we decide if adding extra setup steps, writing more documentation, and adding more code to maintain in pwa-starter is worth it, if we don't analyze the current scenarios we are encountering?

**Comment by adriaanwm** [comment]
1. I'm not comparing our Router to react or preact router, I'm simply saying that if we're using a global store in our projects, and the lib uses context, that's more going on than a router that only uses one or the other. I would be all for separating the Router into something like `BaseRouter` where you have to handle the state things yourself in an intermediary file, and `ContextRouter` or just `Router` that uses context. They could be in separate packages or together. This applies to other packages like `useForm` where already having the state self-contained has been an issue for both me and Cory (and the reason for #132 needing to be created).

2. I guess I meant more than just the API, but the implementation as well. Those functions look simple enough, but they are doing a lot under the hood to get that information from one component to another. I probably should have said "pure functions" instead of just "functions" since stateful functions are worse than pure functions, given all else is equal. At the extreme, a package that just has an API with one function `createApp` sounds great but only at the surface level.

Hooks may be a benefit over Class syntax, but I disagree they're a benefit over HOCs

> Absolutely, it matters

Ok, I agree it does matter whether something occurs or not when weighing out possibilities. But when one solution handles essentially every possibility (the value is in global store, anyone can use it anywhere), whereas the other just assumes only a handful of things are possible, (you don't need route args outside top component) the one that's more versatile should be considered better than the other. And something like accessing the route arg from any component is not a crazy idea.

Another time to use a route arg is if you have a slug in the name and want to render the slug as a title while the actual resource is fetching. Or you need to use it in a form as initialData. Or you are using a third party hook and the id is used.

**Comment by staydecent** [comment]
Ok, that makes much more sense. 

1. This is a nice approach, but I still wonder if it's worth the overhead of more integration code, additional work on the element, and more documentation to write 😅. I just tested out a prop based approach to syncing the router state:

```
export default function App() {
  const syncState = ({ path, route }) => {
    Object.assign(globalState, { currentPath: path, currentRoute: route })
  }
  return (
    <RouteProvider routes={routes} syncState={syncState}>
      <h1>App</h1>
      <Router routes={routes} />
      <Link to="home" activeClass="active" className="home">Take me home</Link>
    </RouteProvider>
  );
}
```

☝️ This works by adding ~2 lines to `router.js` and will call the syncState function *before* the internal `setRoute`, which should sync the state to the global store, before rendering the matched Route Component.

2. I agree with what you are saying 100%, but Context is actually a rather simple Component factory: https://github.com/preactjs/preact/blob/master/src/create-context.js

Hooks vs HoC is interesting. HoC implementation is better, but hooks are "de-facto" in react world now. I think it's ideal for our code to not deviate too much from the "broader react community".

Even we if did decide to favour HoC's, we would, at some point, reach for 3rd party libs that would provide hooks. So, we'd have to be familiar with both anyway. In such case, is it better to only encounter hooks in our projects, or mostly HoC's and sometimes hooks? Is this point worth considering `withRequest` instead of `useRequest`?

I think the more versatile option should be considered more versatile. But that doesn't mean it doesn't come with a cost (more functions exported, more docs needed).

|      | BaseRouter                      | syncState                      |
| ---- | ------------------------------- | ------------------------------ |
| Pros | 100% Global state control       | One prop to learn              |
|      |                                 |                                |
| Cons | Splits Router into 2 exports    | Cannot override internal state |
|      | Requires more docs/code in pwa  |                                |
|      | State values should have checks/guards since they are coming from end-user rather than internally  |                                |

I'm sure I'm missing cons for syncState, but that's how I see the two options currently.

**Comment by staydecent** [comment]
Also interesting is react-navigation and react-router both recommend *against* integrating routing state with Redux...

https://reactnavigation.org/docs/redux-integration#can-i-store-the-navigation-state-in-redux-too
https://reactrouter.com/web/guides/deep-redux-integration

**Comment by staydecent** [comment]
I think this whole discussion can be attributed to the [Use/Reuse Paradox](https://dzone.com/articles/usereuse-paradox).

> A module or service might be highly reusable, but very difficult to use. Likewise, a module or service might be very easy to use, but difficult to reuse. And it’s incredibly difficult to offer both.

Reuse being the ability to make a derivative, and use being embedding the module in a project. I am advocating for ease of use over ease of re-use. Not just to make the project more friendly to the public, but to also reduce the configuration required within pwa-starter. I think pwa-starter should be as close to "just folder structure" as possible.

> Because of this, we should strive to minimize the architectural significance (impact and cost) of change by making our designs as reversible as possible. 

To me, exposing props will be a less intrusive or disruptive change than breaking Router into a Stateless export and a Stateful export. I also am not convinced we should allow internal state to be set from outside of Router, as I tend to agree with react-navigation that it introduces surface area for creating bugs.

The delay in the ability to sync the internal router state to an outside source, is one worth fixing. But I think we should fix it, without compromising the usability of the element.

**Comment by adriaanwm** [comment]
Cool to see the create-context code, simpler than I was picturing, still don't completely understand what's going on in there since I don't know the internals of preact much. But there's still some iffy stuff going on with the exported i variable and this weird use of `some` with an effect https://github.com/preactjs/preact/blob/master/src/create-context.js#L24-L27 .

**Comment by adriaanwm** [comment]
> I tend to agree with react-navigation that it introduces surface area for creating bugs.

Aren't these only the same potential bugs for any use of global store, anyone can change and break it?

So what is our opinion on state?
1. global store for everything
2. mix of global store and component state (with context sometimes)
3. should move away from global store and just use component state with context

My preference has always been 1, because 
- it doesn't limit how / when a developer uses state... more power for developers
- redux devtools make it handy to be able to see all application state
- state is in one place (almost)

Points I can see in 2's favor
- state/context can handle js objects with methods in them, rather than just essentially json in the store
- remove potential bugs by blocking access to internal state
- less setup for elements that use internal state

Am I right to assume your preference is 2?

**Comment by staydecent** [comment]
Indeed, I find the preact coding style to be hard to follow at times. Seems the `i` is so the count increments for all nested providers, previously there was collisions: https://github.com/preactjs/preact/commit/a2317cc5cfd699a5f0007b8545b349fc03263ed9

And not sure why they use some instead of map? This is the only detail I could find: https://github.com/preactjs/preact/commit/3f711099be50bd8e646805f18dd72ad774d1c817

> Aren't these only the same potential bugs for any use of global store, anyone can change and break it?

Yup, so if we don't need to set internal routing state from outside, why introduce it? Router should be the safest/most stable element in our projects.

1. Never been a proponent for 100% global state. Only reach for it if you need to share state, and if you need more than 1 component to alter that state. I think you could look at any of our projects and see a healthy mix of global and local state.

2. Yes, but don't think we need to use Context in project code. Our global state's main purpose is that of Redux's, to have traceable and maintainable state changes coming from multiple components. A nice _extra_ feature of our global state is that we can subscribe to state without relying on props/react hierarchy, making Context redundant in that sense.

3. Definitely not!

Do we have any projects that use 100% global state? I'm pretty sure it's always a balancing act, but maybe it would simplify development for the team, if we only used `useStore` instead of `useState`?

**Comment by adriaanwm** [comment]
I'm not sure I agree with the use/reuse paradox being our underlying issue. Functional tools like `reduce`, `map`, etc. are extremely useful and reusable, we should strive for this imo. `match` which we took from preact router is also very easy to use and reusable. A stateless router is not that hard to set up, and once it is it is just as easy to use as the stateful one. 

Another benefit to go stateless first is that the stateless things are easy to reason about and optimize with things like memoization.

Thinking I should just build the router out so it's easier to compare

**Comment by staydecent** [comment]
How do you reuse reduce? The implementation is hidden? That's solely just "use". Same with `match`.. We are "using" match, it is not exposed in a way that can be "reused". Reused in this sense is the ability to create derivatives.

**Comment by adriaanwm** [comment]
I thought reuse meant being able to use the thing in different situations. Essentially not tightly coupled to a specific use case.

**Comment by staydecent** [comment]
> If we make a software entity highly reusable then it’s likely a lightweight and fine-grained entity. This allows for environmental configuration driven by context and extensibility through well-defined interfaces and extension points. But with this flexibility is additional complexity that makes the entity inherently more difficult to use (flexibility and complexity are another paradox?). 

> In a collective work, software components are put together as is, possibly with some glue code around them. In a derivative work of some component, the original component is modified (that is, its source code is changed). A derivative work can of course be used as part of a collective work.

Exposing a BaseRouter or StatelessRouter is exposing an incomplete component. That is, Router and the related Link, RouteTo, etc. can not work together as they have no mechanism to share state until that is setup. This is optimizing for "reusability" in that more of the internal logic and APIs are exposed increasing the potential for derivative works. Such as creating the GlobalStateRouter we would use in pwa-starter.

I don't think it's worth this flexibility to create derivative works.

> Another benefit to go stateless first is that the stateless things are easy to reason about and optimize with things like memoization.

I don't follow this. Stateless components just means the state has to come in as a prop. This moves the source of state from internal use (Context.Consumer), to outside in the host project. In this case, we would have to add memoization to each component that receives state from props (Link, RouteTo, etc.). But, with context, and managing state internally, we can just add memoization optimizations to the `useRouterState` hook, since all components are deriving state props from that source.

**Comment by adriaanwm** [comment]
State is one of the main sources of bugs and complexity. When the state code is in a package fixing these bugs is especially annoying. If it were entirely up to me, elements would be mostly pure functions, and anywhere state gets introduced would be very explicit and controlled. Separating state from any other code is generally a good practice (the more your functions are pure the better).

**Comment by staydecent** [comment]
> State is one of the main sources of bugs and complexity. 
💯 

> When the state code is in a package fixing these bugs is especially annoying.

Fair enough. But, if there is a bug in a critical component like Router, we can fix and deploy a new version, and update old apps easily. The more code we put in pwa-starter the more we have to c+p fixes to that code on old projects. I don't see why we can't get Router to a stable place in a month, and have minimal updates to it from that point. This is also the reason I'd like to eventually adopt a build/transpile system that requires less configuration than rollup, so pwa-starter is truly just folder structure, and critical updates can be version controlled via npm.

> If it were entirely up to me, elements would be mostly pure functions, and anywhere state gets introduced would be very explicit and controlled. Separating state from any other code is generally a good practice (the more your functions are pure the better).

I agree. But I don't see this as an argument against current implementations. Both Dropdown and Router have all state functions contained in a single hook, respectively. Using context to consume/make available those state functions _is explicit_. `useMappedState` is no more or less pure than context. Both are a subscription service to a value -- which is, they have basically identical side-effects.

*The delay should be fixed though.* And shuold be doable without much fuss. I'll spin up test to confirm for tomorrow.

**Comment by staydecent** [comment]
I actually couldn't reproduce this issue. I converted the User component in `pwa-starter` to get `id` from `useMappedState` instead of the prop passed down via Router.

When the User component  renders, the `id` value retrieved by `useMappedState` is set, and ready to go on initial mount -- it is *not* undefined.

<img width="454" alt="Screen Shot 2020-08-06 at 6 03 56 PM" src="https://user-images.githubusercontent.com/153041/89597732-c89c0a80-d80f-11ea-814d-64c7e23b84a1.png">

But, when I navigate _away_ from User back to the `Users` listing page, useMappedState is updated BEFORE Router renders the new component (Users), because, it is in fact syncing the state before the Router renders the matched component.

In other words: We navigate away from `users/4` and the new route is synced to global state via SyncRouterState, so useMappedState is called again on User component, updating the value of id from "4" to "undefined", then Router renders the newly matched route component `Users`.

This problem would actually exist if Router was 100% controlled by global state. Router, finds a new match, and updates state, any component subscribed to that value re-renders, including the Router component.

So we actually want the matched route component to be rendered by Router, _before_ global state is updated. But if your matched component is trying to get the route args from global state, then you will encounter a race condition anyway.

I imagine this is what react-router and react-navigation are referring to when they say "shooting yourself in the foot".

**Comment by staydecent** [comment]
Another way this would break/result in undefined:

```
// If we render a tree like this, because `/users/:userId` is matched:
<>
  <Header />
  <Router>
    <User id={4} />
  </Router>
</>
```

But, we decide Header will request an avatar to render if on a User page. We will _have_ to check if `currentRoute.args.id` is undefined, because when we navigate away from `/users/4` to `/users` it will be, and yet Header is still rendered.

I believe this is why other routers expose a [Match](https://reach.tech/router/api/Match) component or [hook](https://reactrouter.com/web/api/Hooks/useroutematch).
